blockchain/hedera.go

```diff
@@ -15,63 +17,129 @@ import (
-	//todo see how to use timestamps to request transaction for period after our last request
-	//todo parse result, extract info from memo
-	//todo trigger jobs on chainlink node
-	//todo map somehow task id from memo to a different jobspec? is that possible? what is jobID in hederaSubscription?
-	//todo check if payment was okay
-	//todo create tests - hedera_test.go & also elsewhere where we changed the code + db migration
+	err = viper.ReadInConfig()
+	if err != nil {
+		return
+	}
 
-	//todo see where we lose the /
-	var url = hederaSubscription.endpoint + "/"
+	err = viper.Unmarshal(&hederaConfig)
+	return
+}
 
+func (hSubscr hederaSubscriber) SubscribeToEvents(channel chan<- subscriber.Event, _ store.RuntimeConfig) (subscriber.ISubscription, error) {
 
-	//go tzs.readMessagesWithRetry()
+	hederaSubscription := hederaSubscription{
+		endpoint:  hSubscr.Endpoint,
+		events:    channel,
+		accountId: hSubscr.AccountId,
+		jobid:     hSubscr.JobID,
+	}
 
-	var client = NewClient(url, 5)
+	if len(hederaSubscribersMap[hederaSubscription.accountId]) == 0 {
+		var client = NewClient(hederaSubscription.endpoint, 5)
+		go client.WaitForTransaction(hederaSubscription.accountId)
+	}
 
-	client.GetAccountCreditTransactionsAfterTimestamp("0.0.1943014", 324324234234)
+	addHederaSubscriber(hederaSubscription.accountId, hederaSubscription)
 
 	return hederaSubscription, nil
 }
 
 func (hSubscr hederaSubscriber) Test() error {
+
+	hederaConfig, err := LoadConfig(".env")
+	if err != nil {
+		logger.Error(err)
+		return err
+	} else if hederaConfig.LinkTopicId == "" {
+		return errors.New("LINK Token ID is missing! Please set LINK Token ID to .env configuration file")
+	}
+
+	tokenId = hederaConfig.LinkTopicId
+
+	var client = NewClient(hSubscr.Endpoint, 0)
+	response, err := client.GetAccountByAccountId(hSubscr.AccountId)
+	if err != nil {
+		return errors.New("Error while getting the Account information")
+	}
+
+	if response != nil {
+		if response.Accounts != nil && len(response.Accounts) == 1 {
+			account := response.Accounts[0]
+			if account.Deleted {
+				errorMessage := fmt.Sprintf("Account with ID: %s is deleted", hSubscr.AccountId)
+				return errors.New(errorMessage)
+			} else if account.Balance.Tokens != nil {
+				isAccountHasAssignedLinkToken := false
+				for _, token := range account.Balance.Tokens {
+					if token.TokenId == tokenId {
+						isAccountHasAssignedLinkToken = true
+					}
+				}
+				if !isAccountHasAssignedLinkToken {
+					errorMessage := fmt.Sprintf("Account with ID: %s is not assigned to LINK token with ID: %s", hSubscr.AccountId, tokenId)
+					return errors.New(errorMessage)
+				}
+			} else {
+				errorMessage := fmt.Sprintf("Account with ID: %s is not assigned to LINK token with ID: %s", hSubscr.AccountId, tokenId)
+				return errors.New(errorMessage)
+			}
+		} else {
+			return errors.New("Please check that you have recorded the Hedera's AccountId correctly")
+		}
+	}
```

Comment:
Can we refactor the logic to not be that nested?

Reply:
done